src/main/java/uk/ac/sanger/sccp/stan/service/FileStoreServiceImp.java

```diff
@@ -0,0 +1,118 @@
+    private final StanFileConfig config;
+    private final Clock clock;
+    private final Transactor transactor;
+    private final StanFileRepo fileRepo;
+    private final WorkRepo workRepo;
+
+    @Autowired
+    public FileStoreServiceImp(StanFileConfig config, Clock clock, Transactor transactor,
+                               StanFileRepo fileRepo, WorkRepo workRepo) {
+        this.config = config;
+        this.clock = clock;
+        this.transactor = transactor;
+        this.fileRepo = fileRepo;
+        this.workRepo = workRepo;
+    }
+
+    @Override
+    public StanFile save(User user, MultipartFile fileData, String workNumber) {
+        Work work = workRepo.getByWorkNumber(workNumber);
+
+        String filename = getFilename(fileData);
+        String san = filename.replaceAll("[^a-zA-Z0-9_-]+", "");
+        if (san.isEmpty()) {
+            san = "unnamed";
+        }
+
+        LocalDateTime now = LocalDateTime.now(clock);
+
+        final String savedFilename = now + "_" + san;
+        Path path = Paths.get(config.getDir(), savedFilename);
+
+        try {
+            fileData.transferTo(Paths.get(config.getRoot(), config.getDir(), savedFilename));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        return transactor.transact("updateStanFiles",
+                () -> updateStanFiles(user, filename, work, now, path.toString()));
+    }
+
+    private StanFile updateStanFiles(User user, String originalName, Work work, LocalDateTime now, String storedPath) {
+        deprecateOldFiles(originalName, work.getId(), now);
+        return fileRepo.save(new StanFile(work, user, originalName, storedPath));
+    }
+
+    private String getFilename(MultipartFile file) {
+        String name = file.getOriginalFilename();
+        if (name!=null) {
+            int c = name.lastIndexOf('/');
+            if (c >= 0) {
+                name = name.substring(c+1);
+            }
+        }
+        return (name==null || name.isEmpty() ? "unnamed" : name);
+    }
+
+    @Override
+    public Resource loadResource(StanFile stanFile) {
+        if (!stanFile.isActive()) {
+            throw new IllegalStateException("File is inactive: "+stanFile.getPath());
+        }
+        Path path = Paths.get(config.getRoot(), stanFile.getPath());
+        try {
+            return new UrlResource(path.toUri());
+        } catch (MalformedURLException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @Override
+    public StanFile lookUp(Integer id) {
+        return fileRepo.getById(id);
+    }
+
+    public void deprecateOldFiles(String name, Integer workId, LocalDateTime timestamp) {
+        List<StanFile> oldFiles = fileRepo.findAllActiveByWorkIdAndName(workId, name);
+        if (oldFiles.isEmpty()) {
+            return;
+        }
+        for (StanFile f : oldFiles) {
+            f.setDeprecated(timestamp);
```

Comment:
All previous uploaded files of same name and workid (if any) are updated with current time as deprecated. Do we ever need to use this deprecated time other than to check the deprecated status - if so, do we need to keep the real deprecated times for previously deprecated ones and update it only for last uploaded file (of same name and workid). I don't think this matters much in current context, but just a thought.

Reply:
Only the active (i.e. not deprecated) matching files get deprecated here, so the deprecated timestamp will be correct. For our purposes it could just be a boolean, but for audit and testing purposes it's just as easy to use a timestamp.